// react-interval/lib/ReactInterval.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ReactComponentWithPureRenderMixin = require('react/lib/ReactComponentWithPureRenderMixin');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ReactInterval = _react2.default.createClass({
  displayName: 'ReactInterval',

  propTypes: {
    callback: _react2.default.PropTypes.func.isRequired,
    enabled: _react2.default.PropTypes.bool,
    timeout: _react2.default.PropTypes.number
  },

  getDefaultProps: function getDefaultProps() {
    return {
      enabled: false,
      timeout: 1000
    };
  },
  getInitialState: function getInitialState() {
    return { enabled: this.props.enabled };
  },
  componentDidMount: function componentDidMount() {
    if (this.props.enabled) {
      this.start();
    }
  },
  componentWillReceiveProps: function componentWillReceiveProps(_ref) {
    var enabled = _ref.enabled;

    this.setState({ enabled: enabled });
  },


  shouldComponentUpdate: _ReactComponentWithPureRenderMixin.shouldComponentUpdate,

  componentWillUnmount: function componentWillUnmount() {
    this.stop();
  },
  callback: function callback() {
    this.props.callback();
    this.start();
  },
  start: function start() {
    this.stop();
    this.timer = setTimeout(this.callback, this.props.timeout);
  },
  stop: function stop() {
    clearTimeout(this.timer);
  },
  render: function render() {
    if (this.state.enabled) {
      this.start();
    } else {
      this.stop();
    }
    return false;
  }
});

exports.default = ReactInterval;
